src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java

```diff
@@ -311,7 +363,13 @@ public void run() {
                     int nReady, bytesRead;
 
                     // wait for close or inotify event
-                    nReady = poll(ifd, socketpair[0]);
+                    try {
+                        do {
+                            nReady = poll(ifd, socketpair[0]);
```

Comment:
Let's explain the design first, please, then we can think about the lost wakeup. I wrote the comment actually before I realized that watch keys are not handled.

I understand the hesitation to manage watch keys automatically. What I am missing, though, is an example of idiomatic code where the user actually manages the watch keys during C/R - you only provide a test that checks that C/R fails when the key is left open. Maybe  There I would like to demonstrate that closing the notify service automatically simplifies the code, rather than just adding a close/reopen to the watch key management.